js/walutomat.js

```diff
@@ -0,0 +1,309 @@
+            request['continueFrom'] = since;
+        }
+        if (limit !== undefined) {
+            request['volume'] = limit;
+        }
+        const response = await this.privateGetAccountHistory (this.extend (request, params));
+        return this.parseTrades (response, market, since, limit);
+    }
+
+    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const body = {
+            'submitId': this.uuid (),
+            'pair': market['id'],
+            'price': price,
+            'buySell': side.toUpperCase (),
+            'volumeCurrency': market['baseId'],
+            'otherCurrency': market['quoteId'],
+            'volume': amount,
+        };
+        return await this.privatePostMarketOrders (this.extend (body, params));
+    }
+
+    async cancelOrder (id, symbol = undefined, params = {}) {
+        const request = {
+            'orderId': id,
+        };
+        return await this.privatePostMarketOrdersCloseOrderId (this.extend (request, params));
+    }
+
+    async fetchOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        // TODO: Would be nice to get parametrized API response
+        await this.loadMarkets ();
+        const market = undefined;
+        const orders = await this.privateGetMarketOrders ();
+        return this.parseOrders (orders, market, since, limit);
+    }
+
+    async fetchClosedOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        // TODO: Would be nice to have fetchOrders parametrized to get closed offers
+        return await this.fetchOrders (symbol, since, limit, params);
+    }
+
+    formatOrderBook (orderBookEntry) {
+        return [this.safeFloat (orderBookEntry, 'price'), this.safeFloat (orderBookEntry, 'baseVolume')];
+    }
+
+    parseOrder (order, market = undefined) {
+        const id = this.safeString (order, 'orderId');
+        let symbol = undefined;
+        if ('market' in order) {
+            market = this.safeString (order, 'market');
+            symbol = this.findSymbol (market);
+        }
+        let timestamp = undefined;
+        let datetime = undefined;
+        if ('submitTs' in order) {
+            datetime = this.safeString (order, 'submitTs');
+            timestamp = this.parse8601 (datetime);
+        }
+        let lastTradeTimestamp = undefined;
+        if ('submitTs' in order) {
+            lastTradeTimestamp = timestamp;
+        }
+        let side = undefined;
+        if ('buySell' in order) {
+            side = this.safeString (order, 'buySell').toLowerCase ();
+        }
+        let price = undefined;
+        if ('price' in order) {
+            price = this.safeFloat (order, 'price');
+        }
+        let cost = undefined;
+        if ('feeAmountMax' in order) {
+            cost = this.safeFloat (order, 'feeAmountMax');
```

Comment:
> Done ;)

Ahh... not really ) I mean, for some reason you've wrapped it all in `ifs`, where `this.safeString` and `this.safeFloat` do the necessary check inside, so there was no need in wrapping these conditionals this way, the actual need was to fix the fields purposes, because, you see, the rules for a unified order are:
- `order['cost'] = order['filled'] * order['price']`
- `order['cost'] !== order['fee']['cost']` ←----- in current code the fee is written into the order `cost` field,  and the `fee` field is left undefined, where it is actually defined ) The point is to make each field conform to the documented purpose/meaning from the Manual. This bug was there  in the initial implementation in that other PR from there very beginning.

Reply:
Thanks for your time :) I've updated stuff, but I am not sure if I correctly understand idea of 'filled', is it the amount, which user wants to buy? That's how API response looks like from Walutomat: https://api.walutomat.pl/#lista-aktywnych-ofert-na-gie%C5%82dzie

Reply:
> I am not sure if I correctly understand idea of 'filled', is it the amount, which user wants to buy?

Nope, actually, the order placed on the orderbook has three amounts simultaneously at all times:

- the amount ordered by the user
- the part of the above amount that was ful**filled** (executed) already
- the part of the above amount that is still remaining not fulfilled (in your open offer/order)

The order itself is at all times in a certain state:

- `open` the order has been just opened/placed (amount > 0, filled = 0, remaining = amount - filled = amount)
- `partially filled` the ordered amount was executed by the counter-side, but the order is still not fulfilled yet (amount > 0, filled > 0, filled < amount, remaining = amount - filled)
- `closed` – the order was executed completely (sold our or bought out, amount > 0, filled = amount, remaining = amount - filled = 0)
- `canceled` – the order was canceled by the user before it was completely closed (amount > 0, filled >= 0, remaining = amount - filled)

In the CCXT we do not have a need to distinguish between a partially filled order and an open orders, because the order is still open in both cases, it is the filled amount that changes over time. So, in terms of CCXT, the order can be either `open`, `closed` or `canceled` in general. The `filled` value ranges from 0 to `amount` and designates which part of the order was already closed (as well as which part is still remaining `open – because one can be calculated from another and vice versa).

- https://github.com/ccxt/ccxt/wiki/Manual#how-orders-are-related-to-trades

Let me know if that doesn't answer the question.

Reply:
Ah yeah, now I understand that. Give me some time I will look into the API and get idea how it can be done here :)

Reply:
@piotrmoszkowicz 

> how it can be done here

If it cannot be done or the exchange API implies some incompatible principles – then there's no need to fill the information that you can't have from there, instead we prefer to tell the user the actual truth – namely, if some field wasn't received (missing in the response) and it cannot be calculated with basic arithmetics from the other available fields, then there's no need to fill the missing information and we simply put `undefined` there by default. However, in most cases, exchanges implementing maker/taker model provide two of the three amounts, so you can evaluate the third missing one on the client side inside `parseOrder`, for example.

Reply:
in WT api, `soldAmount` and `boughtAmount` are the closest equivalent to `filled` field. There's also `completion` but it's expressed in percent, so using it to calculate exact filled amount could be inprecise.
I think that `filled`should be ` side === 'buy' ? order.soldAmount : order.boughtAmount`

Re cost, from what I understand it could be:

```js
order['price'] = order['feeRate'] // this is expressed in percent to be deducted
order['cost'] =order['boughtAmount'] * order['price'] // Walutomat is always charging from the bought side, so at any time, transaction cost should boughtAmount*price.
```

does that make sense?

Reply:
> does that make sense?

Yep, you practically correct, one minor clarification: `order['cost'] = filled * price`, so it can't be always `boughtAmount * price` and should be either `boughtAmount * price` or `soldAmount * price`, depending on the side, or am I missing something? My logic behind this is: if `filled` depends on the `side` and the `cost` depends on the `filled`, then `cost` depends on the `side`, right?

Reply:
@jpospychala 

> Not quite, Walutomat always deducts fee from gathered (bought) amount,
for example (assuming EUR/PLN rate 4.0000):

I think, you're falling into your previous confusion again! ))

We're talking of order cost, not fee cost. Those are two different costs. 

The `filled` value is the volume executed in base currency.
The `cost` is the total of `filled` * `price` in quote currency, whereas `fee['cost']` and `fee['currency']` should depend on the order `side`, exactly as the exchange works and as our API works as well. 

Both `costs` are very strictly defined in the Manual in terms which currency it is ) This error was in the code you initially submitted and we are fixing it, so that it corresponds to our unified API spec:

- https://github.com/ccxt/ccxt/wiki/Manual#order-structure

Hope this clarifies it, let me know if not.